### ISSUE TITLE
Memory leak in fit or train function 

### ISSUE THREAD

**Issue by edeforas** [tiny-dnn/tiny-dnn#927]
Hello,

While running this very simple sample:

`#include <iostream>
using namespace std;

#include "tiny_dnn/network.h"
#include "tiny_dnn/layers/fully_connected_layer.h"
#include "tiny_dnn/activations/tanh_layer.h"
using namespace tiny_dnn;

//original code from http://linerocks.blogspot.fr/2017/02/solving-xor-problem-using-tiny-dnn_89.html

int main()
{
    network<sequential> net;

    net << fully_connected_layer(2,3);
    net << tanh_layer();
    net << fully_connected_layer(3,1);
    net << tanh_layer();

    vector<vec_t> input_data  { {0,0}, {0,1}, {1,0}, {1,1} };
    vector<vec_t> desired_out {   {0},   {1}, {1},     {0} };

    size_t batch_size = 1;
    size_t epochs = 50000; //500; big number to show the bug . 
    momentum opt;
    net.fit<mse>(opt, input_data, desired_out, batch_size, epochs);

    double loss = net.get_loss<mse>(input_data, desired_out);
    cout << "mse : " << loss << endl;

    vec_t r00=net.predict(input_data[0]);
    vec_t r01=net.predict(input_data[1]);
    vec_t r10=net.predict(input_data[2]);
    vec_t r11=net.predict(input_data[3]);

    cout << "predict: " << r00[0] << " " << r01[0] << " " << r10[0] << " " << r11[0] << endl;

    return 0;
}`

The memory increase indefinitely:

![mem_leak_xor](https://user-images.githubusercontent.com/16975259/36062441-1deb4df8-0e6d-11e8-8682-813595bbcd1e.png)

There is the same problem with the train function (see issue https://github.com/tiny-dnn/tiny-dnn/issues/736 )
Build under Windows10 with Mingw 492_3 (under QtCreator)

(Also , would you accept a PR for this example (need to ask to author) ? )

Thanks

**Comment by beru** [comment]
> would you accept a PR for this example (need to ask to author) ?

The project maintainers aren't very active lately but if you send a PR that might be of use to someone at least.

If you isolate the bug and if it's apparent. Then reviewing process would be easier.

**Comment by edeforas** [comment]
The bug is still here with mingw 530.
But the bug is gone with vs2017 (x64 build) ! So I will use vs2017 for now.

For the PR, could we create a integration branch that could accept all simple PR, then merge in master when mainteners will be back?

**Comment by edgarriba** [comment]
@edeforas just send a PR against master so that it can be reviewed

**Comment by edeforas** [comment]
I  have a PR branch but have no right to push the features branch ( I wanted to push in feat/add_example_sinus_fit), 
how can I do? Thanks
error is : remote: Permission to tiny-dnn/tiny-dnn.git denied to edeforas.

**Comment by beru** [comment]
@edeforas You can create a new Pull Request.
